controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java

```diff
@@ -126,12 +129,8 @@ public final void removeCapabilityRequirements(OperationContext context, Resourc
         }
 
         private void processCapabilityRequirement(OperationContext context, String attributeName, boolean remove, String... attributeValues) {
-            String dependentName;
-            if (dynamicDependent) {
-                dependentName = RuntimeCapability.buildDynamicCapabilityName(baseDependentName, getDynamicDependentName(context.getCurrentAddress()));
-            } else {
-                dependentName = baseDependentName;
-            }
+            RuntimeCapability<?> dependentCapability = this.getDependentCapability(context);
+            String dependentName = (dependentCapability.isDynamicallyNamed() ? dependentCapability.fromBaseCapability(context.getCurrentAddress()) : dependentCapability).getName();
```

Comment:
The main change is here. 

Before, we were checking if the dependent capability was dynamic (according to the author) to be able to build its name.
Now, we fetch the dependent `RuntimeCapability` according to the current OperationContext MRR and build its name according to the MRR capabilities.

It is similar to the ContextDependencyRecorder implementation (although its `getDependentCapability` impl differs a bit as it checks there is a single registered capability).

My concern is the warning in `ContextDependencyRecorder` javadoc:

> CapabilityReferenceRecorder that determines the dependent capability from the OperationContext. This assumes that the resource registration associated with currently executing step will expose a capability set including one and only one capability. __This recorder cannot be used with attributes associated with resources that do not meet this requirement.__

Your change to the DefaultCapabilityReferenceRecorder generalises the requirement of the ContextDependencyRecorder as the MRR can contain several capabilities and you only use the one corresponding to the `baseDependentName`.

Is my understanding correct?
Is there still a sense to have separate DefaultCapabilityReferenceRecorder and ContextDependencyRecorder as they now both require a Context MRR to work.
We could have a special case for attribute-based capability but that's a special case of your general DefaultCapabilityReferenceRecorder now, right?

Reply:
Your understanding is correct.  In retrospect, DefaultCapabilityReferenceRecorder should probably extend ContextDependencyRecorder, and just override its logic for generating the dependent name (i.e. from a specific baseDependentName).

> Is there still a sense to have separate DefaultCapabilityReferenceRecorder and ContextDependencyRecorder

A valid question.  Personally, I didn't even know that ContextDependencyRecorder existed until this week :) - so I don't know how many people are using it.
ContextDependencyRecorder requires fewer parameters, which is nice, but it can't be used if the resource defines multiple capabilities.
I suspect it is worth having both - but we should at least consolidate the implementations.

Another logical implementation is one that accepts the RuntimeCapability itself (this is how the CapabilityReferenceRecorder in wildfly-clustering-common works), which simplifies the getDependentCapability(...) even further (and without an exception case), but that is probably out of scope for this fix.

> We could have a special case for attribute-based capability

What do you mean by attribute-based capability?

Reply:
Ah - maybe you mean CompositeAttributeDependencyRecorder?

Yeah - to be more universal, I would expect this to be a special case of DefaultCapabilityReferenceRecorder, not ContextDependencyRecorder.